### PR TITLE
feat: implement user lifecycle

### LINKED ISSUE #76: Preserve user in FreeIPA instead of delete
## Question

Is there a way to preserve user (`freeipa_user`) when destroying resource - instead of delete?

### PR DESCRIPTION
**Pull request by infra-monkey**
Closes #76 
This is quite a big pull request, so please can someone do some testing before approving?

Basically, I have implemented a state machine for the user state (staged --> active <-->preserved)
I had to implement a resource modification plan because the lifecycle depends on multiple attributes. But at least illegal transitions will fail early.

Normally, admins will modify the user attibutes only for active state but it also works for staged and preserved state. 
I have not tested all attributes but a few. 

I implemented the tests for the main user lifecycle usecases.
Documentation has been adapted.

The disabled state is a bit wierd. Not possible to set on staged users and not reported by the read function for preserved users. I had to change the way it is handled.

you will see that there are 3 different files for the states of a user that implement read, update and create in slightly different maners. There is also one for the state upgrade. Feel free to propose another way to handle it. I have split the code in multiple files in order for readability, I find it annoying to work on huge files.

Any comment will be appreciated

### CODE REVIEW

**Comment by jagular-spb** [comment]
will take a look

File: docs/resources/user.md [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
 page_title: "freeipa_user Resource - freeipa"
 description: |-
-  FreeIPA User resource
+  FreeIPA User resource. The user lifecycle is managed with the attributes:
+  
+  account_staged
+  account_preserved
+  (defaults to active)
+  An active user can be preserved.
+  A user can bestaged only at the user's creation.
```
Comment (jagular-spb) [inline docs/resources/user.md]:
```suggestion
  A user can be staged only at the user's creation.
```

Comment (infra-monkey) [reply docs/resources/user.md]: Fixed with the next commit

**Comment by infra-monkey** [comment]
Implemented a state upgrader to ensure that the state and additional attributes are correctly set when upgrading from a previous version of the provider

**Comment by jagular-spb** [comment]
> Implemented a state upgrader to ensure that the state and additional attributes are correctly set when upgrading from a previous version of the provider

it works

**Comment by infra-monkey** [comment]
@jagular-spb I've added the transition from preserved to staged now.
Also fixed the import.

**Comment by jagular-spb** [comment]
> This '+' means successful I guess?
Yes
> I don't see transition from preserved to active.
It was checked before, but I`ll do it again
> Have you checked changing other attributes
Nope, good point, will do

**Comment by infra-monkey** [comment]
@jagular-spb did you have time to test the newest commits?

**Comment by jagular-spb** [comment]
> @jagular-spb did you have time to test the newest commits?

I`m back to testing, will report feedback on this week

**Comment by jagular-spb** [comment]
> Can you expand on that? There has been a PR merge because NS records were not allowed in the dns_record resource. #88 are you refering to this? Is there another issue?
Nope, issue was related to absence of NS record type (I thought it was already implemented ...)

> Can you expand on that? I'm not sure to understand
with  3 different variables Ive to get data source for each and merge them, when\if I need to operate with all possible users, 
its not a big deal probably

**Comment by infra-monkey** [comment]
@jagular-spb @RomanButsiy 
I've pushed a commit that implements lifecycle based on a `state` attribute.
I have deprecated `account_disabled`

I have done some testing, including updating from 5.0.3.

It looks okay to me.

jagular, if you have time to test this one, it would be appreciated.

I am also removing the "DO NOT MERGE" title.

**Comment by RomanButsiy** [comment]
@infra-monkey I think this is ready to merge. Everything looks good in my tests. If any edge cases show up later, we can address them in follow-up fixes

**Review by RomanButsiy — APPROVED** [review]
